### ISSUE TITLE
Question regarding Filename on Windows (Ruby bindings...)

### ISSUE THREAD

**Issue by s0nspark** [taglib/taglib#1219]
I am hoping to build the ruby bindings for taglib (which are based on swig) on Windows but have hit a snag... [see this comment](https://github.com/robinst/taglib-ruby/issues/102#issuecomment-747713902) for the details, but the gist of it is that the implementation of Filename in taglib differs on Windows compared to other platforms. As a result the current ruby bindings simply will not build against any recent version of taglib on Windows.

Is this something that could be changed or otherwise addressed in 2.0 to bring the Windows implementation in line with other platforms so we can get Ruby bindings on Windows?

**Comment by ufleisch** [comment]
You should add some conditional code for the Windows build, see [my comment](https://github.com/robinst/taglib-ruby/issues/102#issuecomment-1932851821) in the taglib-ruby issue tracker.

**Comment by ufleisch** [comment]
The removal of the implicit `const char *` conversion of the Windows `FileName` was probably a bad idea. I am toying with the idea of reintroducing this method (and a required `std::string` member) in version 2.0.1, but this would violate binary compatibility only for Windows and only for `FileName`. Opinions?

**Comment by scotchi** [comment]
Because it would introduce a new virtual method?  My sense is that most folks link statically on Windows.  Taking TagLib from just LGPL to LGPL / MPL was mostly to embrace that.

**Comment by ufleisch** [comment]
No, the binary compatibility violation would be "Add new data members to an existing class". `FileName` is now

```
  class TAGLIB_EXPORT FileName
  {
  public:
    FileName(const wchar_t *name);
    FileName(const char *name);

    FileName(const FileName &name);

    operator const wchar_t *() const;

    const std::wstring &wstr() const;

    String toString() const;

  private:
    TAGLIB_MSVC_SUPPRESS_WARNING_NEEDS_TO_HAVE_DLL_INTERFACE
    const std::wstring m_wname;
  };
```

whereas in v1.13.1 it was

```
  class TAGLIB_EXPORT FileName
  {
  public:
    FileName(const wchar_t *name);
    FileName(const char *name);

    FileName(const FileName &name);

    operator const wchar_t *() const;
    operator const char *() const;

    const std::wstring &wstr() const;
    const std::string  &str() const;

    String toString() const;

  private:
    const std::string  m_name;
    const std::wstring m_wname;
  };
```

Adding `const std::string m_name` would be necessary to implement `operator const char *() const`, it must point to some storage. Putting `m_name` and `m_wname` into a `d` pimpl would also change the size of `FileName`. I also think that most people on Windows link statically (so do I), so this change could be acceptable.

**Comment by scotchi** [comment]
Ah, I hadn't realized `FileName` didn't have a d-pointer.  That's kind of annoying.  It seems like I missed adding `FileName` to my class size unit tests last year when I was reviewing all of the public classes.

**Comment by ufleisch** [comment]
The issues which prevented the taglib-ruby bindings from using Taglib 2 were resolved in TagLib 2.0.1. This version is now used in the taglib-2 branch of taglib-ruby, see https://github.com/robinst/taglib-ruby/pull/143#event-12508940687. This branch passes all tests. If you encounter problems with these bindings on Windows, you should report them [there](https://github.com/robinst/taglib-ruby/issues).
